stdlib/string.mli

```diff
@@ -395,6 +395,48 @@ val is_valid_utf_16le : t -> bool
 (** [is_valid_utf_16le b] is [true] if and only if [b] contains valid
     UTF-16LE data. *)
 
+(** {1:spellchecking Spellchecking} *)
+
+val edit_distance : ?limit:int -> t -> t -> int
+(** [edit_distance s0 s1] is the number of single character edits
+    (understood as insertion, deletion, substitution, transposition)
+    that are needed to change [s0] into [s1].
+
+    If [limit] is provided the function returns with [limit] as soon
+    as it was determined that [s0] and [s1] have distance of at least
+    [limit]. This is faster if you have a fixed limit, for example for
+    spellchecking.
+
+    The function assumes the strings are UTF-8 encoded and uses {!Uchar.t}
+    for the notion of character. Decoding errors are replaced by
+    {!Uchar.rep}. Normalizing the strings to
+    {{:https://unicode.org/glossary/#normalization_form_c}NFC} gives
+    better results.
+
+    {b Note.} This implements the simpler Optimal String Alignement (OSA)
+    distance, not the Damerau-Levenshtein distance. With this function
+    ["ca"] and ["abc"] have a distance of 3 not 2.
+
+    @since 5.4
+*)
+
+val spellcheck :
+  ?max_dist:(string -> int) -> string list -> string -> string list
```

Comment:
I am wondering if it would not be better to represent the dictionary as an iterator `(string -> unit) -> unit` rather than a list here.

Indeed, the case when the word belongs to the dictionary feels a bit strange with the current design: we end up traversing the whole list to build a singleton list containing our initial word.

This suggests to me that a common pattern would be to first test if the word belongs exactly to the dictionary (with a string set or a trie) before calling the `spellcheck` function. But then if the dictionary is not a list, it seems better to avoid constructing an intermediary list.

As a small positive side-effect, this would remove the dependency on the `List` module which feels quite accidental.

Reply:
> Indeed, the case when the word belongs to the dictionary feels a bit strange with the current design: we end up traversing the whole list to build a singleton list containing our initial word.

Well usually when you get to use `spellcheck` you already know you have mismatch. That being said @gasche already suggested to have something more flexible than a list so I think I'm ok with making this slightly more complex for users in the name of flexibility. 

> As a small positive side-effect, this would remove the dependency on the `List` module which feels quite accidental.

I will still need `List.rev` :–)

Reply:
I was thinking that the ordering is mostly cosmetics, but it is probably better to preserve a possible alphabetical order from the dictionary.

Reply:
Yes especially because suggestions ("the dictionary") may not in alphabetical order at all but sorted according to some notion of application defined preference.

Reply:
`String.spellcheck`'s dictionaries are now represented by iterators.